Add STM32G0B/C support

chris-durand opened the pull request:
- [x] Enable devices
- [x] Support G0B/C shared interrupt handlers
  - [x] I2C
  - [x] ~~FDCAN~~, unsupported and deactivated on G0B/C for now
  - [x] SPI
- [x] Fix DMA driver (fixed by [device data correction](https://github.com/modm-io/modm-devices/pull/100) in modm-devices)

Not tested yet.

chris-durand committed:
[stm32] Enable STM32G0B/C devices

chris-durand committed:
[stm32] Fix platform startup for STM32G0B/C

chris-durand committed:
[stm32] Add STM32G0B/C support to I2C driver

chris-durand commented on src/modm/platform/i2c/stm32-extended/module.lb:
```diff
@@ -10,6 +11,23 @@
+    irq_re = re.compile("I2C\d(_\d)+")
+    shared_irqs = [v["name"] for v in device.get_driver("core")["vector"]]
+    shared_irqs = [v for v in shared_irqs if irq_re.fullmatch(v)]
+    shared_irq_map = defaultdict(list)
```
```suggestion
    shared_irq_map = {}
```

chris-durand commented on src/modm/platform/i2c/stm32-extended/module.lb:
```diff
@@ -64,12 +90,18 @@ def prepare(module, options):
     return True
 
 def build(env):
+    device = env[":target"]
+
```
```suggestion
```

chris-durand commented on src/modm/platform/i2c/stm32-extended/module.lb:
```diff
@@ -64,12 +90,18 @@ def prepare(module, options):
-    env.copy("i2c_timing_calculator.hpp")
 
+    env.copy("i2c_timing_calculator.hpp")
+    env.template("i2c_shared_irqs.cpp.in")
```
```suggestion
    if len(global_properties["shared_irqs"]) > 0:
        env.template("i2c_shared_irqs.cpp.in")
```

klsc-zeat commented on src/modm/platform/i2c/stm32-extended/i2c_shared_irqs.cpp.in:
```diff
@@ -0,0 +1,25 @@
+/*
```
This file will generated empty if you only use I2C1 -> Is there an option to not generate at all if not needed?

chris-durand replied:
See the comment above. The file won't be generated anymore, if there are no shared interrupts with that change.

klsc-zeat commented on src/modm/platform/i2c/stm32-extended/i2c_master.cpp.in:
```diff
@@ -323,15 +324,18 @@ namespace
 MODM_ISR(I2C{{ id }}_EV)
 %% endif
 {
 	DEBUG_STREAM("\n=== IRQ ===");
```
a few lines late CR1 and ICR are always modified -> in the shared case we should probably check if we have a reason to modify anything:
```
uint16_t isr = I2C{{ id }}->ISR;
if (isr & ?MASK? == 0)
{
    return;
}
```

chris-durand replied:
Thanks for pointing that out. I didn't look too closely at the driver yesterday. We have to return early if no interrupt flag that is enabled is set.

That could work:
```
const uint16_t isr = I2C{{ id }}->ISR;
%% if shared_irq
const auto cr1 = I2C{{ id }}->CR1;
const auto mask = (cr1 & (I2C_CR1_STOPIE | I2C_CR1_RXIE | I2C_CR1_TXIE))
    | ((cr1 & I2C_CR1_TCIE) ? (I2C_ISR_TCR | I2C_ISR_TC) : 0u); 

static_assert(I2C_CR1_STOPIE == I2C_ISR_STOPF);
static_assert(I2C_CR1_TXIE == I2C_ISR_TXIS);
static_assert(I2C_CR1_RXIE == I2C_ISR_RXNE);

if ((isr & mask) == 0) {
    return;
}
%% endif
```

What do you think?

klsc-zeat replied:
I can test this next week -> I need to build a setup that uses both I2C2 and I2C3 first.

klsc-zeat replied:
Ok managed to test your branch today. I2C2 and I2C3 work fine in parallel on my Nucleo-G0B1RE.
![image](https://github.com/modm-io/modm/assets/7805403/64cbbad2-cdb7-4248-bc54-5a7ba176b819)
P.S. pinging 2 HYT221 here

chris-durand committed:
[stm32] Fix DMAMUX max channel calculation

The channel ids were incorrectly sorted as strings and then converted
to int. The conversion must happen first for the correct ordering.
The bug has no practical effect since no device has more than 8
channels.

chris-durand committed:
[ext] Update modm-devices submodule

chris-durand committed:
[stm32] Support STM32G0B/C shared interrupt handler in SPI HAL

chris-durand committed:
[stm32] Do not enable FDCAN driver on STM32G0

FDCAN on STM32G0 is currently not supported due to the shared interrupt
handler.

chris-durand commented:
@klsc-zeat Do you require FDCAN support on STM32G0B/C? I have disabled the driver on those devices for now.

klsc-zeat commented:
> @klsc-zeat Do you require FDCAN support on STM32G0B/C? I have disabled the driver on those devices for now.

Unfortunately yes but only FDCAN1. But i think i will manage that with the examples from the branch. I will post a PR but it will probably take me some time.

salkinium reviewed (approved):
Awesome, this has been on my todo list for a long time!
